bitswap/client/internal/session/session.go

```diff
@@ -75,7 +75,7 @@ type SessionPeerManager interface {
 // ProviderFinder is used to find providers for a given key
 type ProviderFinder interface {
 	// FindProvidersAsync searches for peers that provide the given CID
-	FindProvidersAsync(ctx context.Context, k cid.Cid) <-chan peer.ID
+	FindProvidersAsync(ctx context.Context, k cid.Cid) <-chan peer.AddrInfo
```

Comment:
I think using the `libp2p.ContentDiscovery` interface (i.e. with the `max int`) makes sense.

> Let max be a bitswap.Option and not a ProviderQueryManager one, or call it with -1 or something to signal we want to use the "default".

Why not have bitswap call it with 0 (i.e. give me all of them) and have the ProviderQueryManager handle any "max" setting? The "max" on the FindProvidersAsync interface is about the max providers returned rather than the max providers searched for. The ProviderQueryManager could reasonably choose to search for 100 peers but only return the first 10 that it connected to and still be honoring `FindProvidersAsync(..., 10)`.

Reply:
If ProviderQueryManager implements the ContentDiscovery interface, then it should return an unlimited number of results when called with max == 0, per the requirements of that interface.

We could use <0 as the values that we use the MaxProviders option for (the routing interface does not specify what to do with negative values):

  * `pqm.FindProvidersAsync(ctx, cid, 10)`  -> returns at most 10 providers that we managed to connect to (it may search more)
  * `pqm.FindProvidersAsync(ctx, cid, 0)` -> returns unbounded number of providers
  * `pqm.FindProvidersAsync(ctx, cid, -1)` -> returns at most pqm.maxProviders that we managed to connect to.

Thoughts?

Reply:
We could do this, although is it worth the hassle given at the moment there's only one consumer that likely doesn't need this?

What about this?
- `pqm.FindProvidersAsync(ctx, cid, 10)` returns at most 10 providers that we manage to connect to (it will search for up to pqm.maxProviders which can be set to 0 and ignored)
- `pqm.FindProvidersAsync(ctx, cid, 0)` -> returns unbounded number of providers that we manage to connect to (it will search for up to pqm.maxProviders which can be set to 0 and ignored)
- default pqm.maxProviders is 0

In Bitswap we would default to `pqm.FindProvidersAsync(ctx, cid, 10)` and would set `pqm.maxProviders` to 10. In a followup PR (once we know this behaves well) we'd try setting pqm.maxProviders to 0 by default.

If in the future there's a desire to effectively set `pqm.maxProviders` per query as a parameter we can figure something out like:
1. Just add a new function `FindReachableProvidersAsync(ctx, cid, maxReturnedProviders, maxAttemptedProviders)`
2. Hide the `maxProviders` as a context settable option
3. Use negative numbers
   - `pqm.FindProvidersAsync(ctx, cid, -X)` returns at most X providers that we manage to connect to (it will search for up to X providers). We could then drop the ability to set `pqm.maxProviders` entirely and assert that setting the last parameter >= 0 is the same as the old `pqm.maxProviders=0` (i.e. the default)

Reply:
yeah ok, then we do your proposal

Reply:
fixed in latest commit